Superclass delay

fanfilmu opened the pull request:
I've noticed that when Resque::Retry checks whether or not exception should be retried, it allows thrown exception to be a subclass of any exception from `@retry_exceptions` ([here](https://github.com/lantins/resque-retry/blob/f7088a3d9110f2bffb3899ac872cd6ee6f528961/lib/resque/plugins/retry.rb#L228)), but when fetching delay for such exception, it is required to supply exact exception class ([here](https://github.com/lantins/resque-retry/blob/f7088a3d9110f2bffb3899ac872cd6ee6f528961/lib/resque/plugins/retry.rb#L145)).

This PR changes behaviour so that the closest superclass (or exact exception class, if present `@retry_exceptions`) is used to fetch delay for thrown exception on retry.

I will appreciate any suggestions! :)

jzaleski commented:
I like this change. I'm just concerned about the implementation. Can you think of a way we could achieve the same effect w/o having to `sort` and limit `@retry_exceptions` every time?

fanfilmu commented:
I've had some small luck using `inject`. Here is the benchmark code:

```ruby
require 'benchmark'

class A; end
class B < A; end
class C < B; end
class D < C; end
class Other; end

exceptions = { A => 10, Other => 89 }

select_and_sort = -> do
  relevant_definitions = exceptions.select { |ex| D <= ex }.sort
  relevant_definitions.any? ? relevant_definitions.first[1] : 0
end

inject = -> do
  klass = exceptions.inject(Object) { |acc, ex| ex[0] < acc && ex[0] >= D ? ex[0] : acc }
  exceptions[klass] || 0
end

ancestors = -> do
  D.ancestors.find { |ex| exceptions[ex] } || 0
end

puts "Select & sort"
puts Benchmark.measure { 10_000_000.times { select_and_sort.call } }

puts "Inject"
puts Benchmark.measure { 10_000_000.times { inject.call } }

puts "Ancestors"
puts Benchmark.measure { 10_000_000.times { ancestors.call } }
```

Results:

```
Select & sort
 11.420000   0.000000  11.420000 ( 11.421683)
Inject
  9.270000   0.000000   9.270000 (  9.271558)
Ancestors
 16.760000   0.000000  16.760000 ( 16.765855)
```

It's slightly faster and should consume less memory, since no new arrays are created during `select` and `sort`. What do you think?

The test failed on jruby though :(

lantins commented:
For myself:

- I prefer your original set of changes using `sort` because its easier to grok the code. It's just 'simpler' and I like that.
- There's not much difference in the speed, yeah its a couple of seconds total, but nothing when you look at the 'per call' differences.
- None of this stuff will happen during a 'happy path' through the code anyhow.

fanfilmu commented:
Personally I'd go with first option too.

Execution time for first two approaches depends on the size of `@retry_exceptions` hash. For the third one, `ancestors`, the level of inheritance is more important. For example, same test when there were 9 exceptions in the hash has given these results:

```ruby
Select & sort
 24.360000   0.000000  24.360000 ( 24.363573)
Inject
 23.140000   0.000000  23.140000 ( 23.143807)
Ancestors
 17.750000   0.000000  17.750000 ( 17.753926)
```

So the only sense I make of it is that there is no optimal way and the 'simplest' one might be best :)

jzaleski commented:
I agree w/ @lantins here that it is easier to read and there is value in that!

Maybe we can just split the difference and memoize the result for a `delay` after the first time we do the sort/index? This could be achieved a few ways, either by storing a separate data-structure or updating `@retry_exceptions` ensuring a constant-time lookup the second time.

Happy to discuss this further, but don't want to split hairs.

fanfilmu commented:
Sorry for the delay, I didn't have much time lately to get back to it.

I followed @jzaleski's advice and added assigning found definition to exact exception class. Also, added note about it to method's documentation and test for it.

What do you think?

jzaleski commented:
This seems reasonable to me @fanfilmu. @lantins what do you think?

lantins commented:
Today I learned about `||= begin ... end` block :)

Thanks for your effort @fanfilmu and thanks for the checks and balances @jzaleski
